File: frameworks/runtime-src/Framework/platform/MWIOUtils.cpp
#include "MWIOUtils.h"

#if CC_TARGET_PLATFORM == CC_PLATFORM_IOS
#include "ios/MWIOUtilsStrategyIOS.h"
#elif CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID
#include "android/MWIOUtilsStrategyAndroid.h"
#elif CC_TARGET_PLATFORM == CC_PLATFORM_WIN32
#include "win32/MWIOUtilsStrategyWin32.h"
#endif
#include "cocos2d.h"

using namespace std;
using namespace cocos2d;

MW_FRAMEWORK_BEGIN

MWIOUtils::MWIOUtils()
#if CC_TARGET_PLATFORM == CC_PLATFORM_IOS
: _strategy(new MWIOUtilsStrategyIos())
#elif CC_TARGET_PLATFORM == CC_PLATFORM_ANDROID
: _strategy(new MWIOUtilsStrategyAndroid())
#elif CC_TARGET_PLATFORM == CC_PLATFORM_WIN32
: _strategy(new MWIOUtilsStrategyWin32())
#else
: _strategy(nullptr)
#endif
{
}

MWIOUtils::~MWIOUtils()
{
    CC_SAFE_DELETE(_strategy);
}

string MWIOUtils::resourcePath(const std::string &path)
{
    if (_strategy) {
        return _strategy->resourcePath(path);
    }
    MW_THROW_EXCEPTION(1012);
}

bool MWIOUtils::fileExists(const std::string &path)
{
    return FileUtils::getInstance()->isFileExist(path);
}

bool MWIOUtils::directoryExists(const std::string &path)
{
    return FileUtils::getInstance()->isDirectoryExist(path);
}

bool MWIOUtils::moveFile(const std::string &oldPath, const std::string &newPath)
{
    if (_strategy) {
        return _strategy->moveFile(oldPath, newPath);
    }
    MW_THROW_EXCEPTION(1012);
}

bool MWIOUtils::copyFile(const std::string &oldPath, const std::string &newPath)
{
    if (_strategy) {
        return _strategy->copyFile(oldPath, newPath);
    }
    MW_THROW_EXCEPTION(1012);
}

MWBinaryData *MWIOUtils::getDataFromFile(const std::string &filePath)
{
    if (_strategy) {
        return _strategy->getDataFromFile(filePath);
    }
    MW_THROW_EXCEPTION(1012);
}

bool MWIOUtils::writeDataToFile(const MW_RAW_DATA content, MW_ULONG length, const std::string &filePath, bool isAppend)
{
    if (_strategy) {
        return _strategy->writeDataToFile(content, length, filePath, isAppend);
    }
    MW_THROW_EXCEPTION(1012);
}

bool MWIOUtils::writeDataToFile(mwframework::MWBinaryData *data, const std::string &filePath, bool isAppend)
{
    return this->writeDataToFile(data->getData(), data->getSize(), filePath, isAppend);
}

bool MWIOUtils::removeFile(const std::string &filePath)
{
    if (_strategy) {
        return _strategy->removeFile(filePath);
    }
    MW_THROW_EXCEPTION(1012);
}

bool MWIOUtils::removeDirectory(const std::string &directoryPath)
{
    return cocos2d::FileUtils::getInstance()->removeDirectory(directoryPath);
}

bool MWIOUtils::createFile(const std::string &filePath)
{
    if (_strategy) {
        return _strategy->createFile(filePath);
    }
    MW_THROW_EXCEPTION(1012);
}

bool MWIOUtils::createDirectory(const std::string &directoryPath)
{
    if (_strategy) {
        return _strategy->createDirectory(directoryPath);
    }
    MW_THROW_EXCEPTION(1012);
}

std::string MWIOUtils::splicePath(const std::string &path1, const std::string &path2)
{
    if (path1.size() <= 0) {
        return path1;
    }
    std::string newPath = path1;
    if (path1.at(path1.size() - 1) != '/') {
        newPath.append("/");
    }
    newPath.append(path2);
    
    return newPath;
}

MW_FRAMEWORK_END

File: frameworks/runtime-src/Framework/lua/auto/api/MWIOUtils.lua

--------------------------------
-- @module MWIOUtils
-- @parent_module mw

--------------------------------
-- Check whether the specified file does exist.<br>
-- param path File path to check.<br>
-- return A boolean value.
-- @function [parent=#MWIOUtils] fileExists 
-- @param self
-- @param #string path
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- Copy file from old path to new path.<br>
-- param oldPath Source file path.<br>
-- param newPath New file path.<br>
-- return Operation result.
-- @function [parent=#MWIOUtils] copyFile 
-- @param self
-- @param #string oldPath
-- @param #string newPath
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- @overload self, mw.MWBinaryData, string, bool         
-- @overload self, void, unsigned long, string, bool         
-- @function [parent=#MWIOUtils] writeDataToFile
-- @param self
-- @param #void content
-- @param #unsigned long length
-- @param #string filePath
-- @param #bool isAppend
-- @return bool#bool ret (return value: bool)

--------------------------------
-- Get the raw data from the specified file.<br>
-- param filePath Source file path.<br>
-- return The raw data of the specified file, or nullptr if it failed to read.
-- @function [parent=#MWIOUtils] getDataFromFile 
-- @param self
-- @param #string filePath
-- @return MWBinaryData#MWBinaryData ret (return value: mw.MWBinaryData)
        
--------------------------------
-- 
-- @function [parent=#MWIOUtils] splicePath 
-- @param self
-- @param #string path1
-- @param #string path2
-- @return string#string ret (return value: string)
        
--------------------------------
-- Create a new file.<br>
-- param filePath The new file path.<br>
-- return Operation result.
-- @function [parent=#MWIOUtils] createFile 
-- @param self
-- @param #string filePath
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- Get the absolute path to the relative resource path.<br>
-- param path Relative path to the resource folder.<br>
-- return The absolute path.
-- @function [parent=#MWIOUtils] resourcePath 
-- @param self
-- @param #string path
-- @return string#string ret (return value: string)
        
--------------------------------
-- Check whether the specified directory does exist.<br>
-- param path Directory path to check.<br>
-- return A boolean value.
-- @function [parent=#MWIOUtils] directoryExists 
-- @param self
-- @param #string path
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- Remove the specified file.<br>
-- param filePath Destination file path.<br>
-- return Operation result.
-- @function [parent=#MWIOUtils] removeFile 
-- @param self
-- @param #string filePath
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- Move file from old path to new path.<br>
-- param oldPath Source file path.<br>
-- param newPath New file path.<br>
-- return Operation result.
-- @function [parent=#MWIOUtils] moveFile 
-- @param self
-- @param #string oldPath
-- @param #string newPath
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- Remove the specified directory.<br>
-- param filePath Destination file path.<br>
-- return Operation result.
-- @function [parent=#MWIOUtils] removeDirectory 
-- @param self
-- @param #string directoryPath
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- Create a new directory.<br>
-- param directoryPath The new directory path.<br>
-- return Operation result.
-- @function [parent=#MWIOUtils] createDirectory 
-- @param self
-- @param #string directoryPath
-- @return bool#bool ret (return value: bool)
        
--------------------------------
-- 
-- @function [parent=#MWIOUtils] getInstance 
-- @param self
-- @return MWIOUtils#MWIOUtils ret (return value: mw.MWIOUtils)
        
return nil
